ext/rugged/rugged_repo.c

```diff
@@ -1504,11 +1504,12 @@ static VALUE flags_to_rb(unsigned int flags)
 
 static int rugged__status_cb(const char *path, unsigned int flags, void *payload)
 {
-	rb_funcall((VALUE)payload, rb_intern("call"), 2,
-		rb_str_new_utf8(path), flags_to_rb(flags)
-	);
+	int exception = GIT_OK;
```

Comment:
This is confusingly named, since it mostly should not be, if would be less confusing if this were a literal 0 since that's what `exception` will be if everything goes well. We can then check whether `exception` is non-0 and then return whatever we feel is appropriate. Let's not be stingy with the number of variables.

Reply:
Looking at this more in context I see now why it's so awkward as this is expecting a libgit2 return code but we need to preserve the exception id or whatever `rb_protect` returns.

On the one hand we should switch to a better iterator via `git_status_list` but keeping with this kind of loop, this is what `GIT_EUSER` is for. We guarantee that we pass this along and that the library won't generate this. You can then use the payload (or whatever other mechanism you have) to store whatever information and if you see `GIT_EUSER` returned, then you _know_ that this was generated by you and you can look up the information in the payload that you passed to the foreach function.